staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/controller_reconcile.go

```diff
@@ -208,19 +319,58 @@ func (c *celAdmissionController) reconcilePolicyBinding(namespace, name string,
 	}
 
 	// Remove compiled template for old binding
-	info.validator.Store(nil)
+	info.validator = nil
 	info.lastReconciledValue = binding
 	return nil
 }
 
-func (c *celAdmissionController) reconcileParams(namespace, name string, params *unstructured.Unstructured) error {
+func (c *policyController) reconcileParams(namespace, name string, params *unstructured.Unstructured) error {
 	// Do nothing.
 	// When we add informational type checking we will need to compile in the
 	// reconcile loops instead of lazily so we can add compiler errors / type
 	// checker errors to the status of the resources.
 	return nil
 }
 
+// Fetches the latest set of policy data or recalculates it if it has changed
+// since it was last fetched
+func (c *policyController) latestPolicyData() []policyData {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.cachedPolicies != nil {
```

Comment:
I'd consider starting with a read lock that checks if this is nil, then going in to a write lock to guard against refreshPolicies being called twice.

This might happen if the triggers for refreshing policies get more complex.

There may be other ways of handling that though (e.g. a workqueue-like structure for refreshing policies may be a better, if more complex-to-write model)

Reply:
> I'd consider starting with a read lock that checks if this is nil, then going in to a write lock to guard against refreshPolicies being called twice.
> This might happen if the triggers for refreshing policies get more complex.

👍🏼 added.

> There may be other ways of handling that though (e.g. a workqueue-like structure for refreshing policies may be a better, if more complex-to-write model)

Given this is an internal class and that it is designed for its one goroutine consumer to poll it in a loop, I think I am happy with the simple design with the RLock mitigation you proposed. I think you are right that if the triggers became more complicated something else might be needed at that time.